keps/sig-etcd/4647-cluster-feature-gate/README.md

```diff
@@ -0,0 +1,597 @@
+
+```go
+DefaultEtcdServerFeatureGates := map[Feature]VersionedSpecs {
+		featureA: VersionedSpecs{
+			{Version: mustParseVersion("3.6"), Default: false, PreRelease: Beta},
+			{Version: mustParseVersion("3.7"), Default: true, PreRelease: GA},
+		},
+		featureB: VersionedSpecs{
+			{Version: mustParseVersion("3.7"), Default: false, PreRelease: Alpha},
+		},
+}
+DefaultEtcdClusterFeatureGates := map[Feature]VersionedSpecs {
+		featureC: VersionedSpecs{
+			{Version: mustParseVersion("3.7"), Default: false, PreRelease: Beta},
+		},
+		featureD: VersionedSpecs{
+			{Version: mustParseVersion("3.6"), Default: false, PreRelease: Alpha},
+			{Version: mustParseVersion("3.7"), Default: true, PreRelease: Deprecated},
+		}
+}
+```
+
+### Set the Feature Gates
+To set the correct user expectation, we will add a new flag `--cluster-feature-gates` (+`config-file` configuration support) similar to the server `--feature-gates` to set the values of cluster feature gates as expected by the local server. The server will use this flag value to set the `proposed_cluster_params` sent to the cluster. The final values of the cluster feature gate are actually decided through raft consensus.
+
+We do not support dynamically changing the feature gates when the server is running.
+
+### Consensus Algorithm
+
+To guarantee consistent value of if a feature is enabled in the whole cluster, the leader would decide if the feature is enabled for all cluster members, and propagate the information to all members. To make our algorithm also extensible to other cluster level parameters, we will change our discussions of the consensus algorithm from cluster feature gates to a more general concept of cluster parameters.
+
+1. When a cluster starts, similar to `ClusterVersion`, the `ClusterParams` would be set to nil, which means all cluster level features are disabled, and other parameters are unknown.
+
+1. When the there is an update of `ClusterVersion`, the `ClusterParams` would be reset to the default values at the new `ClusterVersion`, because the `ClusterParams` values at the previous `ClusterVersion` may no longer be applicable to the new version, thus invalid. The new default `ClusterParams` is set as part of the `ClusterVersionSetRequest`. At the same time, when a server applies updates of the `ClusterVersion`, it will set the `ClusterParams` to the new default `ClusterParams` sent by the leader, all the old `proposed_cluster_params` of all the members stored on the server will be reset to nil to wait for the new values to be refreshed. Each field in `ClusterParams` will have a `versionpb.etcd_version_field` annotation to indicate when it is introduced, and the default should be empty if `ClusterVersion` is smaller than the version when the filed is introduced.
```

Comment:
We should respect flags provided during initial cluster bootstrap. If user explicitly disabled a feature because it's bugged we should not enable at all even if it is default. Based on that I don't think we should initially set `ClusterParams` to default based on `ClusterVersion`.